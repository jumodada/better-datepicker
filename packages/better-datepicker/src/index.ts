export {
  createDatePicker,
  defaultOptions,
  locale,
} from '@better-datepicker/core'
