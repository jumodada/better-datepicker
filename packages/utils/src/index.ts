export * from './typeOf';

export * from './types/index';
