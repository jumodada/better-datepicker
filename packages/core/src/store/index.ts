import { State, States } from '../types/store'
import { initState } from './state'
import { objectKeys } from '../utils/objectKeys'

const Store = (function () {
  let id = 0
  const states: States = {}

  function getState() {
    return states[id]
  }

  function removeState(id: number) {
    delete states[id]
  }

  function createState(options: Partial<State>): State {
    return (states[id] = initState(options, id))
  }

  function runDestroyed(state: State) {
    if (state) {
      state.destroyed?.()
    }
  }

  function destroyed(partialStates?: State[]) {
    if (partialStates) {
      partialStates.forEach((state) => runDestroyed(states[state.id]))
    } else {
      objectKeys(states).forEach((key) => runDestroyed(states[key]))
      id = 0
    }
  }

  return { createState, getState, removeState, destroyed }
})()

export const { createState, getState, removeState, destroyed } = Store
