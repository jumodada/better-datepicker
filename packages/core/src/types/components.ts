import { CellsData, RangeType, State } from './store'
import {
  CreateElement,
  CreateElementOptions,
  CreateElementRequiredOptions,
  eventHandler,
  Fn,
  PartialAtLeastOne,
} from './utils'
import { _EventListener } from './utils'
import { PickersMap, DateType, MonthType, YearType } from './store'

export interface createMonthOrYearComponentsFunction {
  (state: State, t: keyof RangeType): PartialAtLeastOne<CreateElementOptions>
}

export type PickerConfigMap = PickersMap<{
  children: (CreateElementRequiredOptions | CreateElement)[]
  watch: (Fn | Fn[])[]
}>

type EventMethod = (this: State) => void

export type DayEvent = DateType<EventMethod, _EventListener[]>

export type MonthEvent = MonthType<EventMethod, _EventListener[]>

export type YearEvent = YearType<EventMethod, _EventListener[]>

export interface RangeClickEvent {
  complete: {
    plt: keyof RangeType
    status: 'selecting'
  }
  selecting: {
    plt: keyof RangeType
    status: 'complete'
  }
}

interface CreateComponentsOptions {
  listener: (child: CellsData, state: State) => eventHandler
  children: (
    idx: number
  ) => (PartialAtLeastOne<CreateElementOptions> | CreateElement)[]
}

export interface ComponentsType<C = never> {
  month: C
  year: C
}

export type CreateMonthOrYearComponentsOptions =
  ComponentsType<CreateComponentsOptions>
