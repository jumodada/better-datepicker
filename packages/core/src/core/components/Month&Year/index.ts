import { visible } from '../../../utils/element'
import { canIShow } from '../utils'
import { RangeType } from '../../../types/store'
import {
  ComponentsType,
  createMonthOrYearComponentsFunction,
  CreateMonthOrYearComponentsOptions,
} from '../../../types/components'
import { monthEvent, yearEvent } from './event'
import { CreateElementPartOptions } from '../../../types/utils'
import _for from '../../../utils/for'
import { getTenRange } from '../../../utils/date'

const rows = 3
const cols = 4

export function YM(
  componentName: keyof ComponentsType = 'month'
): createMonthOrYearComponentsFunction {
  return function (t: keyof RangeType = 'start'): CreateElementPartOptions {
    const components: CreateMonthOrYearComponentsOptions = {
      month: {
        listener: (child, state) =>
          monthEvent(child)[state.options.type as 'date'],
        children: (idx) => [{ text: this.locale.months[idx] }],
      },
      year: {
        listener: (child, state) =>
          yearEvent(child)[state.options.type as 'date'],
        children: (idx: number) => [
          {
            text() {
              return String(getTenRange(this[t].year)[idx])
            },
          },
        ],
      },
    }
    const { children, listener } = components[componentName]
    function tBody(): CreateElementPartOptions {
      return {
        children: tr(),
        name: 'tbody',
      }
    }

    const tr = (): CreateElementPartOptions[] => {
      return _for((rc) => {
        return {
          name: 'tr',
          children: td(rc),
        }
      }, rows)
    }

    const td = (rc: number): CreateElementPartOptions[] => {
      return _for((cc) => {
        const idx = rc * cols + cc
        const child = this[t][('_' + componentName) as '_month'][idx]

        return {
          name: 'td',
          event: listener(child, this),
          children: children(idx),
          class: () => child.status,
        }
      }, cols)
    }

    return {
      name: 'table',
      children: [tBody()],
      class: [componentName],
      $style: canIShow(() => visible(this.page === componentName)),
    }
  }
}

export const Month = YM()

export function endMonth(): CreateElementPartOptions {
  return Month.call(this, 'end')
}

export const Year = YM('year')

export function endYear(): CreateElementPartOptions {
  return Year.call(this, 'end')
}
