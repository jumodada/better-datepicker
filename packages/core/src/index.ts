import BD from './core'
import Options from './types/options'
import {BetterPickerInstance} from "./types/core"
import {changeOpt} from "./core/util/default-options"
import {pickerLocale} from "./store/modules/options"
import {destroyed} from "./store"

export function createDatePicker(el: HTMLInputElement, options?: Partial<Options>): BetterPickerInstance {
    const picker = BD()
    return <BetterPickerInstance>picker(el, options)
}

export const defaultOptions = changeOpt

export const locale = pickerLocale

export const destroy = destroyed
