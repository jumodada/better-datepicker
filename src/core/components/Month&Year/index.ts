import {visible} from '../../../utils/element'
import {utilStyle} from '../utils'
import {State, RangeType} from '../../../types/store'
import {
    ComponentsType,
    createMonthOrYearComponentsFunction,
    CreateMonthOrYearComponentsOptions,
} from "../../../types/components"
import {monthEvent, yearEvent} from "./event"
import {CreateElementPartOptions} from "../../../types/utils"
import _for from "../../../utils/for"
import {getTenRange} from "../../../utils/date"

let type: keyof RangeType = 'start'
const rows = 3
const cols = 4

const components: CreateMonthOrYearComponentsOptions = {
    month: {
        listener: (child, state) => monthEvent(child)[state.type as 'date'],
        children: (idx, months) => [
            {text: months[idx]}
        ]
    },
    year: {
        listener: (child, state) => yearEvent(child)[state.type as 'date'],
        children: (idx: number) => [{
            text: {
                key: {
                    name: type,
                    childKey: ['year']
                },
                cb: y => String(getTenRange(y)[idx])
            }
        }]
    }
}

export function YM(componentName: keyof ComponentsType = 'month'): createMonthOrYearComponentsFunction {
    const {children, listener} = components[componentName]
    return function (state: State, t: keyof RangeType = 'start'): CreateElementPartOptions {
        type = t

        function tBody(): CreateElementPartOptions {
            return {
                children: tr(),
                name: 'tbody'
            }
        }

        function tr(): CreateElementPartOptions[] {
            return _for((rc) => {
                return {
                    name: 'tr',
                    children: td(rc)
                }
            }, rows)
        }

        function td(rc: number): CreateElementPartOptions[] {
            return _for((cc) => {
                const idx = rc * cols + cc
                const child = state[type][('_' + componentName) as '_month'][idx]
                return {
                    name: 'td',
                    event: {
                        listener: listener(child, state),
                        arg: child
                    },
                    children: children(idx, state.locale.months),
                    class: {
                        key: {
                            name: ['status'],
                            child
                        },
                        cb: (val: string) => val
                    },
                }
            }, cols)
        }

        return {
            name: 'table',
            children: [tBody()],
            class: [componentName],
            style: utilStyle,
            $style: {
                display: {
                    key: ['page'],
                    cb: (page: string) => visible(page === componentName)
                }
            }
        }
    }
}

export const Month = YM()

export function endMonth(state: State): CreateElementPartOptions {
    return Month(state, 'end')
}

export const Year = YM('year')

export function endYear(state: State): CreateElementPartOptions {
    return Year(state, 'end')
}
