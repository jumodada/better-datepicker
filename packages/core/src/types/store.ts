import { Style } from './utils'
import { Callback } from './core'

export type ComponentStatus =
  | 'none'
  | 'pre'
  | 'next'
  | 'selected'
  | 'inRange'
  | 'inRangeWeek'
  | 'range-start'
  | 'range-end'
  | 'range-start range-end'
  | 'today'
  | 'disabled'
  | 'weekStart'
  | 'weekEnd'

export interface CellsData {
  status: ComponentStatus
  date: {
    year: number
    month: number
    day: number
  }
}

export interface MonthType<T = unknown, U = unknown> {
  date: T
  'date-range': T
  'date-week': T
  month: T
  'month-range': T | U
}

export interface YearType<T = unknown, U = unknown> {
  date: T
  'date-range': T
  'date-week': T
  year: T
  month: T
  'year-range': T | U
}

export interface DateType<T = unknown, U = unknown> {
  date: T
  'date-week': T
  'date-range': T | U
}

export interface PickersMap<T = unknown, U = any>
  extends DateType,
    MonthType,
    YearType {
  date: T
  'date-range': T | U
  'date-week': T
  month: T
  'month-range': T | U
  year: T
  'year-range': T | U
}

export type Placement = 'top' | 'left' | 'bottom' | 'right'

export interface DateData {
  year: number
  month: number
  day: number
  date: CellsData['date']
  _date: CellsData[]
  _month: CellsData[]
  _year: CellsData[]
}

export interface LocaleConfig {
  name: string
  weekdays: string[]
  months: string[]
  weekStart: 0 | 1 | 2 | 3 | 4 | 5 | 6
  yearFormat: string
  weekFormat: string
  yearStart: number
}

export type getRangeStatus = 'complete' | 'selecting'

export interface RangeType<T = null> {
  start: T
  end: T
}

export interface Range extends RangeType<CellsData['date'] | null> {
  status: getRangeStatus
  range: CellsData['date'][]
}

export type mode = 'date' | 'year' | 'month'

export interface State {
  reference: HTMLInputElement | null
  popover: HTMLElement | null
  onChange?: Callback
  visible: boolean
  locale: LocaleConfig
  destroyed?: Callback
  update?: Callback
  create?: Callback
  hoverSelected: Range
  start: DateData
  end: DateData
  today: Date
  date: string | string[] | null
  placement: Placement
  placeholder: string
  style: Style
  classes: string[]
  type: keyof PickersMap
  isRange: boolean
  isWeek: boolean
  unlinkPanels: boolean
  offset: number
  zIndex: number
  mode: mode
  format: string
  insertTo: 'body' | 'parent'
  binding: boolean
  disabledDate?: ((date: Date) => boolean) | null
  disabled: boolean
  themeColor: string
  rangeBgColor: string
  tdColor: string
  thColor: string
}

export interface States {
    [key: number]: State
}
