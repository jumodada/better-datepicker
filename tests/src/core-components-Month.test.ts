import {createState} from "../../src/store"
import defaultOptions from "../../src/core/util/default-options"
import {endMonth} from "../../src/core/components/Month&Year"

const state = createState(defaultOptions())
describe('endMonth', () => {
    it('should return HTMLElement', () => {
        expect(endMonth(state)).toMatchSnapshot()
    })
})
