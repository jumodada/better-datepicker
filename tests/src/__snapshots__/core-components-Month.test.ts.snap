// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`endMonth should return HTMLElement 1`] = `
Object {
  "$style": Object {
    "display": Object {
      "cb": [Function],
      "key": Array [
        "page",
      ],
    },
  },
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "text": "Jan",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
            Object {
              "children": Array [
                Object {
                  "text": "Feb",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
            Object {
              "children": Array [
                Object {
                  "text": "Mar",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
            Object {
              "children": Array [
                Object {
                  "text": "Apr",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
          ],
          "name": "tr",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "text": "May",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
            Object {
              "children": Array [
                Object {
                  "text": "Jun",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
            Object {
              "children": Array [
                Object {
                  "text": "Jul",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
            Object {
              "children": Array [
                Object {
                  "text": "Aug",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
          ],
          "name": "tr",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "text": "Sep",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
            Object {
              "children": Array [
                Object {
                  "text": "Oct",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
            Object {
              "children": Array [
                Object {
                  "text": "Nov",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
            Object {
              "children": Array [
                Object {
                  "text": "Dec",
                },
              ],
              "class": Object {
                "cb": [Function],
                "key": Object {
                  "child": Object {
                    "date": "",
                    "status": "",
                  },
                  "name": Array [
                    "status",
                  ],
                },
              },
              "event": Object {
                "arg": Object {
                  "date": "",
                  "status": "",
                },
                "listener": [Function],
              },
              "name": "td",
            },
          ],
          "name": "tr",
        },
      ],
      "name": "tbody",
    },
  ],
  "class": Array [
    "month",
  ],
  "name": "table",
  "style": Object {
    "padding": "5px 20px",
    "text-align": "center",
    "width": "320px",
  },
}
`;
